### DISCUSSION TITLE
How to find which process is active in a calculation?

### DISCUSSION THREAD

**Discussion by mnabideltares** [KratosMultiphysics/Kratos#11410] [General]
Hi,
I am trying to couple heat (temperature) with water-pressure using COSIM. The water-pressure affects the temperature distribution by means of water discharge (convection). The temperature affects the water-pressure by means of water density and viscosity.

So, when I solve the pressure, I need to understand whether the heat is active or not. So, I added a condition to the pressure element:

        if (Geom[0].SolutionStepsDataHas(TEMPERATURE)) {
            this->UpdateWaterProperties(Variables, GPoint);    //Recalculates the water density and viscosity, based on temperature
        }

However, I found this condition is always True even if I don't involve the heat in the calculation. It leads to recalculating the density and viscosity based on temperature=0. The same problem is also at the side of heat element.

            if (Geom[0].SolutionStepsDataHas(WATER_PRESSURE))  {
                this->UpdateWaterProperties(Variables, GPoint);
                this->CalculateDischargeVector(Variables);
            }

For problems with Temperature alone, without pressure, the condition "SolutionStepsDataHas(WATER_PRESSURE)" still gives True.

After debugging, I came to conclusion that, this condition gives True for all registered variables in Kratos, regardless they are involved or not. For example,  "SolutionStepsDataHas(ROTATION)" is still true, while we haven't involved any rotation.

Is this function is intended to give True for all registered variables?

If well, is there any fuction to tell us whether there is solution of a specific variable or not?

NB: In COSIM, we solve the elements seperately, so for Pressure element, there is no DOF for temperature, hence HasDOF is not applicable. While we have temperature values on the nodes, which is exchanged by COSIM.

Thank you in advance for your help.
Regards

**Comment by mnabideltares** [comment]
Update: This comes from "geomechanics_solver.py" where all variables are defined in "def AddVariables(self):"
Setting conditional statement leads to define only the required variables. However, using COSIM, the thermal solver gives the following error, as the pressure is not initialized for the thermal solver. 

    File "d:\Kratos\10372-HeatTransfer-temp-branch-cosim\bin\Debug\KratosMultiphysics\CoSimulationApplication\coupling_interface_data.py", line 76, in _RaiseException
    raise Exception('CouplingInterfaceData "{}" for ModelPart "{}" of SolverWrapper "{}":\n{}'.format(self.name, self.model_part_name, self.solver_name, err_msg))
    Exception: CouplingInterfaceData "WaterPressure" for ModelPart "PorousDomain.Soil" of SolverWrapper "ThermalSolver":
    "WATER_PRESSURE" is missing as SolutionStepVariable in ModelPart "PorousDomain.Soil"

The same occurs also for the pressure solver as the temperature can't be found.

Any hint?

Thank you

**Comment by AlejandroCornejo** [comment]
Hi @mnabideltares , whenever you have nodes that belong to different solvers, they should have the variables needed for each one. In case you coupling two solvers, you must make sure that the shared nodes initialize all the required variables

**Reply by mnabideltares** [reply]
Hi @AlejandroCornejo ,
Thank you for reply.
Then I need to add a new keyword to indicate that the solvers are coupled, then use this keyword as condition for initialization of variables.
    "thermal_pressure_coupled":true  //in both pressure and thermal input for 2-way coupling

**Comment by mnabideltares** [comment]
It is done by adding a keyword in the json file.
